[AMDGPU][CodeGen] Update support (soffset + offset) s_buffer_load's

Acim-Maravic opened the pull request:
getBaseWithConstantOffset() is used for scalar and non-scalar buffer loads. Diffrence between s_load and load instruction is that s_load instruction extends 32-bit offset to 64-bits, so a 32-bit (address + offset) should not cause unsigned 32-bit integer wraparound, because it performs addition in 64-bits.

arsenm commented on llvm/lib/Target/AMDGPU/AMDGPUGlobalISelUtils.cpp:
```diff
@@ -33,6 +33,11 @@ AMDGPU::getBaseWithConstantOffset(MachineRegisterInfo &MRI, Register Reg,
   if (Def->getOpcode() == TargetOpcode::G_ADD) {
+    // A 32-bit (address + offset) should not cause unsigned 32-bit integer
+    // wraparound, because s_load instructions perform the addition in 64 bits.
+    if (!Def->getFlag(MachineInstr::NoUWrap) &&
```
Do we have the same handling in the DAG path? I thought we didn't?

Acim-Maravic replied:
Yes, in SDAG in case of s_buffer_load there is also check for NUW flag in AMDGPUDAGToDAGISel::SelectSMRDBaseOffset. In SDAG they have diffrent functions for scalar and non-scalar buffer_loads.

arsenm commented on llvm/lib/Target/AMDGPU/AMDGPUGlobalISelUtils.cpp:
```diff
@@ -33,6 +33,11 @@ AMDGPU::getBaseWithConstantOffset(MachineRegisterInfo &MRI, Register Reg,
+    // A 32-bit (address + offset) should not cause unsigned 32-bit integer
+    // wraparound, because s_load instructions perform the addition in 64 bits.
+    if (!Def->getFlag(MachineInstr::NoUWrap) &&
+        MRI.getType(Reg).getScalarSizeInBits() == 32 && checkNUW)
```
Capitalize checkNUW and move it first to be next to the !nuw part

kosarev commented on llvm/lib/Target/AMDGPU/AMDGPUGlobalISelUtils.cpp:
```diff
@@ -33,6 +33,11 @@ AMDGPU::getBaseWithConstantOffset(MachineRegisterInfo &MRI, Register Reg,
+    // A 32-bit (address + offset) should not cause unsigned 32-bit integer
+    // wraparound, because s_load instructions perform the addition in 64 bits.
+    if (!Def->getFlag(MachineInstr::NoUWrap) &&
+        MRI.getType(Reg).getScalarSizeInBits() == 32 && checkNUW)
```
In which cases `Reg` is not a 32-bit register? Do we/can we have some tests that cover such cases?

Acim-Maravic replied:
I do not think that there is case where Reg is not 32-bit register. I can make test for that...

kosarev replied:
The check replaced with an assert LGTM.

Acim-Maravic committed:
[AMDGPU][CodeGen] Update support (soffset + offset) s_buffer_load's

getBaseWithConstantOffset() is used for scalar and non-scalar buffer
loads. Diffrence between s_load and load instruction is that s_load
instruction extends 32-bit offset to 64-bits, so a 32-bit
(address + offset) should not cause unsigned 32-bit integer wraparound,
because it performs addition in 64-bits.

Acim-Maravic commented:
@kosarev @arsenm ping

mbrkusanin reviewed (approved):
LGTM

jayfoad commented:
+ @ruiling
